components/user/Tabs/SettingsContent.tsx

```diff
@@ -1,59 +1,237 @@
+  authedUser: API.ApiUserMetadata | null
+  toast: ReturnType<typeof useQueriedToast>
+  reloadUser: () => void
+  onTabChange: (tab: TabType) => unknown
 }
 
-const getSocialChoice = (
-  authedUser: API.ApiUserMetadata
-): { choice: string; value: string } => {
-  if (authedUser.discord) {
-    return { choice: 'discord', value: authedUser.discord }
-  } else if (authedUser.telegram) {
-    return { choice: 'telegram', value: authedUser.telegram }
-  }
-  return { choice: '', value: '' }
+const EDITABLE_FIELDS = {
+  email: { ...FIELDS.email, validation: () => true, touched: true },
+  graffiti: { ...FIELDS.graffiti },
+  country: { ...FIELDS.country, useDefault: false },
+  discord: {
+    ...FIELDS.social,
+    required: false,
+    options: undefined,
+    id: 'discord',
+    label: 'Discord',
+    placeholder: 'Your Discord username',
+    isRadioed: false,
+    validation: () => true,
+  },
+  telegram: {
+    ...FIELDS.social,
+    required: false,
+    options: undefined,
+    id: 'telegram',
+    label: 'Telegram',
+    placeholder: 'Your Telegram username',
+    isRadioed: false,
+    validation: () => true,
+  },
 }
 
-export default function SettingsContent({ authedUser }: Props) {
-  const $email = useField(FIELDS.email)
-  const $graffiti = useField(FIELDS.graffiti)
-  const $social = useField(FIELDS.social)
-  const $country = useField(FIELDS.country)
+export default function SettingsContent({
+  anyBlocksMined,
+  user,
+  authedUser,
+  toast,
+  reloadUser,
+  onTabChange,
+}: Props) {
+  const [$error, $setError] = useState<string>(UNSET)
+  const [$userData, $setUserData] = useState<API.ApiUserMetadata | null>(
+    authedUser
+  )
+
+  const {
+    email: _email = UNSET,
+    graffiti: _graffiti = UNSET,
+    discord: _discord = UNSET,
+    telegram: _telegram = UNSET,
+    country_code: _cc = UNSET,
+  } = $userData || {}
+  const $graffiti = useField({
+    ...EDITABLE_FIELDS.graffiti,
+    defaultValue: _graffiti,
+  })
+
+  const $email = useField({
+    ...EDITABLE_FIELDS.email,
+    defaultValue: _email,
+  })
+  const $discord = useField({
+    ...EDITABLE_FIELDS.discord,
+    defaultValue: _discord,
+    touched: !!_discord,
+  })
+  const $telegram = useField({
+    ...EDITABLE_FIELDS.telegram,
+    defaultValue: _telegram,
+    touched: !!_telegram,
+  })
 
-  const { choice: socialChoice, value: socialValue } =
-    getSocialChoice(authedUser)
+  const $country = useField({
+    ...EDITABLE_FIELDS.country,
+    defaultValue: _cc,
+  })
+  const testInvalid = useCallback(() => {
+    const invalid =
+      !$email?.valid ||
+      !$graffiti?.valid ||
+      !$discord?.valid ||
+      !$telegram?.valid ||
+      !$country?.valid
+    if (invalid) {
+      $setError('Please correct the invalid fields below')
+      scrollUp()
+    } else {
+      $setError(UNSET)
+    }
+    return invalid
+  }, [$email, $graffiti, $telegram, $discord, $country])
+  const update = useCallback(async () => {
```

Comment:
These effects/callbacks should have line breaks between them so they're a bit easier to distinguish from each other.